crates/gitbutler-git/src/bin/askpass/windows-pipe.rs

```diff
@@ -0,0 +1,159 @@
+use std::ffi::OsString;
+use std::io::{self, Read, Write};
+use std::os::windows::ffi::OsStrExt;
+use std::os::windows::io::{AsRawHandle, FromRawHandle, IntoRawHandle, RawHandle};
+use std::path::Path;
+use windows::core::PWSTR;
+use windows::Win32::Foundation::{
+    CloseHandle, DuplicateHandle, BOOL, DUPLICATE_SAME_ACCESS, ERROR_PIPE_NOT_CONNECTED,
+    GENERIC_READ, GENERIC_WRITE, HANDLE, WIN32_ERROR,
+};
+use windows::Win32::Storage::FileSystem::{
+    CreateFileW, FlushFileBuffers, ReadFile, WriteFile, FILE_ATTRIBUTE_NORMAL, FILE_SHARE_READ,
+    FILE_SHARE_WRITE, OPEN_EXISTING,
+};
+use windows::Win32::System::Pipes::{WaitNamedPipeW, NMPWAIT_USE_DEFAULT_WAIT};
+use windows::Win32::System::Threading::GetCurrentProcess;
+
+#[derive(Debug)]
+struct Handle {
+    inner: HANDLE,
+}
+
+unsafe impl Send for Handle {}
+unsafe impl Sync for Handle {}
+
+impl Handle {
+    fn try_clone(&self) -> io::Result<Handle> {
+        let mut new_handle = HANDLE::default();
+        let current_ps_handle: HANDLE = unsafe { GetCurrentProcess() };
+        let res = unsafe {
+            DuplicateHandle(
+                current_ps_handle,
+                self.inner,
+                current_ps_handle,
+                &mut new_handle as *mut HANDLE,
+                0,
+                BOOL(1),
+                DUPLICATE_SAME_ACCESS,
+            )
+        };
+        match res {
+            Ok(_) => Ok(Handle { inner: new_handle }),
+            Err(err) => Err(io::Error::from_raw_os_error(err.code().0)),
+        }
+    }
+}
+
+impl Drop for Handle {
+    fn drop(&mut self) {
+        let _ = unsafe { CloseHandle(self.inner) };
+    }
+}
+
+#[derive(Debug)]
+pub struct Pipe {
+    handle: Handle,
+}
+
+impl Pipe {
+    pub fn connect(path: &Path) -> io::Result<Pipe> {
+        let mut os_str: OsString = path.as_os_str().into();
+        os_str.push("\x00");
+        let mut wide_path: Vec<u16> = os_str.encode_wide().collect();
+
+        let pwstr_path = PWSTR(wide_path.as_mut_ptr());
```

Comment:
I feel like there's got to be some helpers to make these backflips less annoying!